### PR TITLE
Java agent : JVM shutdown hook for debug logging

### PR DESCRIPTION
**Pull request by jabubake**
Logs heap info, thread dump to Cloud Logging on JVM shutdown

### CODE REVIEW

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/ShutdownReporter.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+  public static boolean getStopFlag() {
+    return stopFlag;
+  }
+
+  private void getStackTraceLog(StringBuilder sb, ThreadInfo threadInfo) {
+    sb.append(threadInfo.toString());
+    //ThreadInfo toString is limited to top 8 elements of stack trace
+    StackTraceElement[] ste = threadInfo.getStackTrace();
+    if (ste.length > 8) {
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/ShutdownReporter.java]: 8 should be a static final (constant) value. It's currently a magic number.

It's not clear to me if it has to always be 8 or if it was chosen arbitrarily.

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/TimeoutTracker.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package com.google.cloud.runtimes.monitoring.shutdown;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/TimeoutTracker.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/TimeoutTracker.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+import com.google.cloud.runtimes.monitoring.shutdown.log.ILogging;
+
+public class TimeoutTracker extends Thread {
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/TimeoutTracker.java]: Public classes need javadocs.

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/TimeoutTracker.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+  private int timeoutInMillis;
+  private ILogging logging;
+
+  /**
+   * Notify shutdown hook thread that time out period has elapsed.
+   * Note : Does not terminate shutdown logging thread.
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/TimeoutTracker.java]: Nit: The note should be in a new paragraph. Use a `<p>` tag.

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/config/AgentConfig.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
+package com.google.cloud.runtimes.monitoring.shutdown.config;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/config/AgentConfig.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/config/AgentConfig.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class AgentConfig {
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/config/AgentConfig.java]: javadoc

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/config/LogConfig.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
+package com.google.cloud.runtimes.monitoring.shutdown.config;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/config/LogConfig.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/config/LogConfig.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+import com.google.cloud.logging.Severity;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class LogConfig {
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/config/LogConfig.java]: javadoc

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/config/SysEnvironment.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package com.google.cloud.runtimes.monitoring.shutdown.config;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/config/SysEnvironment.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/config/SysEnvironment.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+public class SysEnvironment {
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/config/SysEnvironment.java]: javadoc.

I guess this class is mostly so you can mock out the environment variables in unit tests?

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/CloudLogging.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
+package com.google.cloud.runtimes.monitoring.shutdown.log;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/CloudLogging.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/CloudLogging.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+import com.google.cloud.runtimes.monitoring.shutdown.config.LogConfig;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+
+public class CloudLogging implements ILogging {
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/CloudLogging.java]: javadoc

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/CloudLogging.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+  private final String logName;
+  private final MonitoredResource monitoredResource;
+  private final String logPrefix;
+  private final Map<String, String> logLabels;
+  private final LogBuffer logBuffer;
+  private final Severity severity;
+
+  /**
+   * Instantiate logging via StackDriver
+   * Maintains a buffered list of logEntry events till flush is called.
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/CloudLogging.java]: nit: add `<p>` tag.

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/ILogging.java [hunk 14]
```diff
@@ -0,0 +1,12 @@
+package com.google.cloud.runtimes.monitoring.shutdown.log;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/ILogging.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/ILogging.java [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+
+public interface ILogging {
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/ILogging.java]: Javadoc.

Also, I would recommend adding javadoc for each method to make it clear how to implement this interface.

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/LogBuffer.java [hunk 16]
```diff
@@ -0,0 +1,82 @@
+package com.google.cloud.runtimes.monitoring.shutdown.log;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/LogBuffer.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/LogBuffer.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class LogBuffer {
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/LogBuffer.java]: javadoc

File: monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/LogBuffer.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+   * @param maxBufferSize max buffer size
+   */
+  public LogBuffer(int maxBufferSize) {
+    this.buffer = new StringBuilder(maxBufferSize);
+    this.maxBufferSize = maxBufferSize;
+    this.chunks = new ArrayList<>();
+  }
+
+  /**
+   * Convert string to be logged into chunks maintaining new line boundaries. Text in a single line
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/java/com/google/cloud/runtimes/monitoring/shutdown/log/LogBuffer.java]: Start a new paragraph after the first sentence.

File: monitoring/shutdown-javaagent/src/main/resources/scripts/format-env-shutdown-javaagent.sh [hunk 19]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/main/resources/scripts/format-env-shutdown-javaagent.sh]: Missing license header.

File: monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/AgentTest.java [hunk 20]
```diff
@@ -0,0 +1,32 @@
+package com.google.cloud.runtimes.monitoring.shutdown;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/AgentTest.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/AgentTest.java [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+import java.lang.reflect.Field;
+import java.util.Map;
+
+public class AgentTest {
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/AgentTest.java]: FYI: test classes don't require javadocs, so I won't be commenting on the other test classes for that. :-)

File: monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/AgentTest.java [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+
+
+  @Test
+  public void hasRegisteredShutdownHook() throws NoSuchFieldException, ClassNotFoundException,
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/AgentTest.java]: Could you add a comment for `// Arrange`, `// Act`, and `// Assert` blocks? I'm having trouble identifying what is the method under test and what is set up / clean up.

File: monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/ShutdownReporterTest.java [hunk 23]
```diff
@@ -0,0 +1,65 @@
+package com.google.cloud.runtimes.monitoring.shutdown;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/ShutdownReporterTest.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/config/AgentConfigTest.java [hunk 24]
```diff
@@ -0,0 +1,60 @@
+package com.google.cloud.runtimes.monitoring.shutdown.config;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/config/AgentConfigTest.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/ShutdownReporterTest.java [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+import com.google.cloud.runtimes.monitoring.shutdown.log.TestLogging;
+
+import org.junit.Test;
+
+public class ShutdownReporterTest {
+
+  private TestLogging logging = new TestLogging();
+
+  @Test
+  public void doesNothingIfHeapInfoOrStackTraceNotEnabled() {
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/ShutdownReporterTest.java]: Good test name. I like that it says the conditions for the test and the expected result.

Could you name the other tests in this class in a similar manner? I would like to see what the expected result is so we can verify that the tests are correct.

File: monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/config/TestSysEnvironment.java [hunk 26]
```diff
@@ -0,0 +1,38 @@
+package com.google.cloud.runtimes.monitoring.shutdown.config;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/config/TestSysEnvironment.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/log/CloudLoggingTest.java [hunk 27]
```diff
@@ -0,0 +1,55 @@
+package com.google.cloud.runtimes.monitoring.shutdown.log;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/log/CloudLoggingTest.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/log/LogBufferTest.java [hunk 28]
```diff
@@ -0,0 +1,37 @@
+package com.google.cloud.runtimes.monitoring.shutdown.log;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/log/LogBufferTest.java]: Missing license header.

File: monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/log/TestLogging.java [hunk 29]
```diff
@@ -0,0 +1,39 @@
+package com.google.cloud.runtimes.monitoring.shutdown.log;
```
Comment (tswast) [inline monitoring/shutdown-javaagent/src/test/java/com/google/cloud/runtimes/monitoring/shutdown/log/TestLogging.java]: Missing license header.

**Comment by lesv** [comment]
I'd prefer to wait until the next update to the PR before I review.  I agree w/ Tim's asks.  Could you verify that we have all of your files (24), as I'm seeing a lot of partial files, and want to make sure when I expand that everything is there.

**Comment by lesv** [comment]
We may also need to get the branch updated.

**Comment by jabubake** [comment]
Les, https://github.com/GoogleCloudPlatform/openjdk-runtime/pull/42/files : lists the 24 files.

**Comment by meltsufin** [comment]
Closing in favor of SIGQUIT/jstack approach.

**Pull request closed without merging**